### PR TITLE
Rabbit integration

### PR DESCRIPTION
**Pull request by pepeArenas**
Contains configuration to connect to rabbitmq
Docker container for rabbitmq
Insertion and retrieve products

### CODE REVIEW

File: management-core/docker-compose.yml [hunk 1]
```diff
@@ -12,4 +12,16 @@ services:
...
     container_name: mysqldbserver
+  rabbit1:
+    image: rabbitmq
+    environment:
+      RABBITMQ_ERLANG_COOKIE: SWQOKODSQALRPCLNMEQG
+      RABBITMQ_DEFAULT_USER: admin
+      RABBITMQ_DEFAULT_PASS: admin
+      RABBITMQ_DEFAULT_VHOST: /
+    ports:
+    - 15672:15672
```
Comment (jahv) [inline management-core/docker-compose.yml]: Use different port

File: management-core/src/main/java/com/ns/task/service/ProductCoreServiceImpl.java [hunk 2]
```diff
@@ -17,18 +17,27 @@
...
 @Service
 public class ProductCoreServiceImpl implements ProductService {
-
-    private final ModelMapper mapper;
-    private final ProductRepository repository;
+    private ModelMapper mapper;
+    private ProductRepository repository;
     private static final Logger logger = LogManager.getLogger();
 
+    public ProductCoreServiceImpl() {
```
Comment (jahv) [inline management-core/src/main/java/com/ns/task/service/ProductCoreServiceImpl.java]: Do you need the empty constructor?

File: management-core/pom.xml [hunk 3]
```diff
@@ -10,7 +10,6 @@
     <modelVersion>4.0.0</modelVersion>
     <artifactId>management-core</artifactId>
     <properties>
-        <mysql-connector-java.version>8.0.12</mysql-connector-java.version>
```
Comment (jahv) [inline management-core/pom.xml:13]: Why did you remove this?

File: management-core/src/main/java/com/ns/task/service/ProductCoreServiceImpl.java [hunk 4]
```diff
@@ -17,18 +17,27 @@
...
+    }
 
     @Autowired
     public ProductCoreServiceImpl(ModelMapper mapper, ProductRepository repository) {
         this.mapper = mapper;
         this.repository = repository;
     }
 
+    @RabbitListener(queues = "q.management.get")
```
Comment (jahv) [inline management-core/src/main/java/com/ns/task/service/ProductCoreServiceImpl.java]: Use the constant string defined in the configuration class

File: management-core/src/main/java/com/ns/task/service/ProductCoreServiceImpl.java [hunk 5]
```diff
@@ -38,15 +47,23 @@ public ProductCoreServiceImpl(ModelMapper mapper, ProductRepository repository)
...
+        return insertProduct(product);
+    }
+
     @Override
     public ProductDTO insertProduct(ProductDTO product) {
         ProductEntity productEntity = convertToEntity(product);
         try {
             logger.debug("Sending data to DB {}", productEntity);
             productEntity = repository.saveProduct(productEntity);
+            product.setMessage("Name and model of the service exists already");
```
Comment (jahv) [inline management-core/src/main/java/com/ns/task/service/ProductCoreServiceImpl.java]: Why setting this message here?

File: management-core/src/main/java/com/ns/task/service/ProductCoreServiceImpl.java [hunk 6]
```diff
@@ -38,15 +47,23 @@ public ProductCoreServiceImpl(ModelMapper mapper, ProductRepository repository)
...
             logger.debug("Getting persisted data from insert to DB {}", productEntity);
         } catch (DataIntegrityViolationException exception) {
-            throw new ProductManagementException("Name and model of the product exists already", exception);
+            product.setMessage("Name and model already exists");
+            return product;
```
Comment (jahv) [inline management-core/src/main/java/com/ns/task/service/ProductCoreServiceImpl.java]: Is not a good idea to return inside a catch, it's preferable to keep the flow, in this case, you can set the message in the product entity and return at the end.

File: management-core/src/main/java/com/ns/task/service/ProductCoreServiceImpl.java [hunk 7]
```diff
@@ -38,15 +47,23 @@ public ProductCoreServiceImpl(ModelMapper mapper, ProductRepository repository)
...
         }
         return convertToDto(productEntity);
```
Comment (jahv) [inline management-core/src/main/java/com/ns/task/service/ProductCoreServiceImpl.java:67]: At the beginning of the method you convert a DTO to an Entity, and at the end, you convert an Entity to a DTO, it's kind of weird and I think it's unnecessary these mappings.

File: management-core/src/test/java/com/ns/task/services/ProductCoreServiceImpTest.java [hunk 8]
```diff
@@ -81,14 +80,17 @@ public void insertProduct() throws DataIntegrityViolationException {
...
     }
 
-    @Test(expected = ProductManagementException.class)
+    @Test
     public void shouldThrowAnExceptionWhenInsertDuplicateProduct() {
         ProductDTO productDTO = new ProductDTO();
         productDTO.setName("SCREWDRIVER");
         productDTO.setModel("S090");
         productDTO.setPrice(new BigDecimal("12.20"));
+        productDTO.setMessage("Name and model already exists");
```
Comment (jahv) [inline management-core/src/test/java/com/ns/task/services/ProductCoreServiceImpTest.java]: You can omit this and just validate the product contains the message.
Use constants

File: management-core/src/main/java/com/ns/task/service/ProductCoreServiceImpl.java [hunk 9]
```diff
@@ -38,15 +47,23 @@ public ProductCoreServiceImpl(ModelMapper mapper, ProductRepository repository)
...
     public ProductDTO insertProduct(ProductDTO product) {
         ProductEntity productEntity = convertToEntity(product);
         try {
             logger.debug("Sending data to DB {}", productEntity);
             productEntity = repository.saveProduct(productEntity);
+            product.setMessage("Name and model of the service exists already");
             logger.debug("Getting persisted data from insert to DB {}", productEntity);
         } catch (DataIntegrityViolationException exception) {
-            throw new ProductManagementException("Name and model of the product exists already", exception);
+            product.setMessage("Name and model already exists");
```
Comment (jahv) [inline management-core/src/main/java/com/ns/task/service/ProductCoreServiceImpl.java]: Use constants for the messages

File: management-web/src/main/java/com/ns/task/service/ProductServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+    @Autowired
+    public ProductServiceImpl(RabbitTemplate rabbitTemplate) {
+        this.rabbitTemplate = rabbitTemplate;
+    }
+
+    @Override
+    public List<ProductDTO> getProducts() {
+        return (List<ProductDTO>) rabbitTemplate.convertSendAndReceive(RabbitConfig.EXCHANGE_GET,
+                RabbitConfig.ROUTING_KEY_GET,
+                "getAllProducts");
```
Comment (jahv) [inline management-web/src/main/java/com/ns/task/service/ProductServiceImpl.java]: Use constants

**Review by jahv — CHANGES_REQUESTED** [review]

File: management-core/src/main/java/com/ns/task/service/ProductCoreServiceImpl.java (hunk 7 above)
Comment (pepeArenas) [reply management-core/src/main/java/com/ns/task/service/ProductCoreServiceImpl.java:67]: This is because my interface service interface is receiving a DTO then I have to go to repository which expect a Entity, this is my current flow

ProductService(DTO)---Interface common between web and core--- ProductService(DTO)-ProductServiceImpl(DTO)-ProducRepository(Entity)—Database

I do know if this is still valid but I knew that is not good to expose the database properties directly so that is why I receive DTO then pass to Entity and return the DTO, what do you think about?

File: management-core/src/main/java/com/ns/task/service/ProductCoreServiceImpl.java (hunk 2 above)
Comment (pepeArenas) [reply management-core/src/main/java/com/ns/task/service/ProductCoreServiceImpl.java]: Yes you are right, do you know a tool that give some advices for this unused code? because I just installed checkstyle but it not giving me this kind of information

File: management-core/pom.xml (hunk 3 above)
Comment (pepeArenas) [reply management-core/pom.xml:13]: When I introduce my custom exception, I was returning to commons DataIntegrity.. which is in that package so by mistake I was doing some test in order to share that import with common project and I just forget to move back this property, because at the end I not need this reference anymore but in the core module

File: management-core/src/main/java/com/ns/task/service/ProductCoreServiceImpl.java (hunk 4 above)
Comment (pepeArenas) [reply management-core/src/main/java/com/ns/task/service/ProductCoreServiceImpl.java]: Fixed, but I have some doubt if I can externalize this one, because  need a constant type as same as the messages for validation, so I fix here but how I have the spring cloud integration in another branch maybe it will change this ono

File: management-core/src/main/java/com/ns/task/service/ProductCoreServiceImpl.java (hunk 5 above)
Comment (pepeArenas) [reply management-core/src/main/java/com/ns/task/service/ProductCoreServiceImpl.java]: You are right, my mistake sorry I'll be more careful

File: management-core/src/main/java/com/ns/task/service/ProductCoreServiceImpl.java (hunk 6 above)
Comment (pepeArenas) [reply management-core/src/main/java/com/ns/task/service/ProductCoreServiceImpl.java]: Fixed, I not will return from a catch again

File: management-core/src/test/java/com/ns/task/services/ProductCoreServiceImpTest.java (hunk 8 above)
Comment (pepeArenas) [reply management-core/src/test/java/com/ns/task/services/ProductCoreServiceImpTest.java]: Fixed

File: management-core/src/main/java/com/ns/task/service/ProductCoreServiceImpl.java (hunk 9 above)
Comment (pepeArenas) [reply management-core/src/main/java/com/ns/task/service/ProductCoreServiceImpl.java]: Fixed

File: management-web/src/main/java/com/ns/task/service/ProductServiceImpl.java (hunk 10 above)
Comment (pepeArenas) [reply management-web/src/main/java/com/ns/task/service/ProductServiceImpl.java]: Fixed

File: management-core/docker-compose.yml (hunk 1 above)
Comment (pepeArenas) [reply management-core/docker-compose.yml]: fixed

**Review by jahv — APPROVED** [review]
